src/clib/pio_nc4.c

```diff
@@ -1023,49 +844,42 @@ int PIOc_get_chunk_cache(int iotype, int io_rank, size_t *sizep,
 #ifdef _NETCDF
 #ifdef _NETCDF4
     case PIO_IOTYPE_NETCDF4P:
-	ierr = nc_get_chunk_cache(sizep, nelemsp, preemptionp);
+	ierr = nc_get_chunk_cache((size_t *)sizep, (size_t *)nelemsp, preemptionp);
 	break;
     case PIO_IOTYPE_NETCDF4C:
-	if (!io_rank)
-	    ierr = nc_get_chunk_cache(sizep, nelemsp, preemptionp);
+	if (!ios->io_rank)
+	    ierr = nc_get_chunk_cache((size_t *)sizep, (size_t *)nelemsp, preemptionp);
 	break;
 #endif
     case PIO_IOTYPE_NETCDF:
-	return PIO_ENOTNC4;
+	ierr = PIO_ENOTNC4;
 	break;
 #endif
 #ifdef _PNETCDF
     case PIO_IOTYPE_PNETCDF:
-	return PIO_ENOTNC4;
+	ierr = PIO_ENOTNC4;
 	break;
 #endif
     default:
 	ierr = iotype_error(iotype,__FILE__,__LINE__);
     }
 
-    /* Allocate an error string if needed. */
-    if (ierr != PIO_NOERR)
-    {
-	errstr = (char *) malloc((strlen(__FILE__) + 20)* sizeof(char));
-	sprintf(errstr,"in file %s",__FILE__);
-    }
 
     /* Check for netCDF error. */
-    /* ierr = check_netcdf(file, ierr, errstr,__LINE__);*/
-
-    /* Free the error string if it was allocated. */
-    if (errstr != NULL)
-	free(errstr);
-
-    /* if (sizep) */
-    /* 	if ((ierr = MPI_Bcast(sizep, 1, MPI_UNSIGNED_LONG, ios->ioroot, ios->my_comm))) */
-    /* 	    return PIO_EIO; */
-    /* if (nelemsp) */
-    /* 	if ((ierr = MPI_Bcast(nelemsp, 1, MPI_UNSIGNED_LONG, ios->ioroot, ios->my_comm))) */
-    /* 	    return PIO_EIO; */
-    /* if (preemptionp) */
-    /* 	if ((ierr = MPI_Bcast(preemptionp, 1, MPI_FLOAT, ios->ioroot, ios->my_comm))) */
-    /* 	    return PIO_EIO; */
+    if (ierr)
```

Comment:
OK, good question!

I have changed code to handle this better, and also added a test which
passes in a deliberately bad value for preemption to ensure that all tasks
get the resulting error code.

Please take a look at the changes and let me know what you thiink.

Thanks,
Ed

On Tue, Mar 1, 2016 at 11:04 AM, jedwards4b <notifications@github.com>
wrote:

> In src/clib/pio_nc4.c
> <https://github.com/NCAR/ParallelIO/pull/59#discussion_r54588679>:
>
> > -    /* ierr = check_netcdf(file, ierr, errstr,__LINE__);*/
> > -
> > -    /* Free the error string if it was allocated. */
> > -    if (errstr != NULL)
> > -	free(errstr);
> > -
> > -    /* if (sizep) */
> > -    /* 	if ((ierr = MPI_Bcast(sizep, 1, MPI_UNSIGNED_LONG, ios->ioroot, ios->my_comm))) */
> > -    /* 	    return PIO_EIO; */
> > -    /* if (nelemsp) */
> > -    /* 	if ((ierr = MPI_Bcast(nelemsp, 1, MPI_UNSIGNED_LONG, ios->ioroot, ios->my_comm))) */
> > -    /* 	    return PIO_EIO; */
> > -    /* if (preemptionp) */
> > -    /* 	if ((ierr = MPI_Bcast(preemptionp, 1, MPI_FLOAT, ios->ioroot, ios->my_comm))) */
> > -    /* 	    return PIO_EIO; */
> > +    if (ierr)
>
> Suppose you are using NETCDF4C and the call from io-task 0 returns an
> error, all other tasks have ierr=PIO_NOERR,
> what happens here?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NCAR/ParallelIO/pull/59/files#r54588679>.
>